File: src/SOP8/Base/Account/Base.Account.BLL/Process/SessionManager.cs
using System;
using System.Collections;
using Base.Account.IBLL.Response;
using Base.Account.IBLL.Models;
using Base.Model.Account;
using Base.DAL;
using dnsDapperDBUtil.DataAccessLayer.IDAL;
using Response.Resource;

namespace Base.Account.BLL.Process
{
    using Resource;

    class SessionManager
    {
        private const int SessionSeconds = 300;

        public static LoginResult CheckLoginSession(IDataManager dataManager, int nUserNo, string strSessionKey)
        {
            LoginResult result = new LoginResult();

            try
            {
                string strErrorMessage;
                string strCondition = string.Format("a.{0} = {1}", Session.Fields.user_sn, nUserNo);

                JoinManager joinManager = new JoinManager(dataManager);
                ArrayList arrDatas = joinManager.JoinSessionUserGrade(strCondition, out strErrorMessage);

                if (arrDatas == null)
                {
                    System.Diagnostics.Trace.WriteLine(strErrorMessage);

                    result.Success = false;
                    result.Message = ID.Get<ErrorMessage>("failToReadSession").Value();
                    return result;
                }
                else if (arrDatas.Count == 0)
                {
                    result.Success = false;
                    result.Message = ID.Get<ErrorMessage>("noneSession").Value();
                    return result;
                }
                else
                {
                    int nDataCount = arrDatas.Count;

                    for (int i=0;i<nDataCount-2;i+=3)
                    {
                        if (arrDatas[i] is Session && arrDatas[i + 1] is User && arrDatas[i + 2] is Grade)
                        {
                            Session session = arrDatas[i] as Session;
                            User user = arrDatas[i + 1] as User;
                            Grade grade = arrDatas[i + 2] as Grade;

                            if (session.session_key == strSessionKey)
                            {
                                // 자동 로그인 확인 여부
                                if (session.atmc_login_yn == false)
                                {   // 자동 로그인이 아니라면
                                    // 마지막 Session 업데이트 시간 체크(현 시간으로부터 300초 이내인지)
                                    DateTime dtSession = session.updt_de;
                                    DateTime dtNow = DateTime.Now;

                                    TimeSpan diffTime = dtNow - dtSession;
                                    double dSecond = diffTime.TotalSeconds;

                                    if (dSecond > SessionSeconds)
                                    {
                                        result.Success = false;
                                        result.Message = ID.Get<ErrorMessage>("logoutSession").Value();
                                        return result;
                                    }
                                    else
                                    {
                                        session.updt_de = DateTime.Now;

                                        if (dataManager.GetUpdate().Update<Session>(session, null, out strErrorMessage) == false)
                                        {
                                            result.Success = false;
                                            result.Message = ID.Get<ErrorMessage>("failToUpdateSession").Value();
                                        }
                                    }
                                }

                                result.Success = true;
                                result.Message = ID.Get<ErrorMessage>("validSession").Value();
                                result.User = ApplicationUser.MakeUser(user, grade, session.session_key);

                                return result;
                            }
                            else
                            {
                                result.Success = false;
                                result.Message = ID.Get<ErrorMessage>("anotherOneLogin").Value();
                                return result;
                            }
                        }
                    }

                    result.Success = false;
                    result.Message = ID.Get<ErrorMessage>("noneSession").Value();
                }
            }
            catch (Exception e)
            {
                result.Success = false;
                result.Message = e.Message;
            }

            return result;
        }
    }
}

File: src/SOP8/Egovframework/Base/base-account-bll/src/main/java/egovframework/base/account/bll/process/SessionManager.java
package egovframework.base.account.bll.process;

import egovframework.base.account.ibll.models.ApplicationUser;
import egovframework.base.account.ibll.response.LoginResult;
import egovframework.base.dal.IDatabase;
import egovframework.base.dal.account.SessionDAO;
import egovframework.base.dal.join.JoinDAO;
import egovframework.base.generic.Holder;
import egovframework.base.model.account.Grade;
import egovframework.base.model.account.Session;
import egovframework.base.model.account.User;
import egovframework.base.account.bll.resource.ErrorMessage;
import egovframework.base.response.resource.ID;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.List;

public class SessionManager {

    private static final int SESSION_SECONDS = 300;

    public static LoginResult checkLoginSession(int userNo, String sessionKey, JoinDAO joinDao, SessionDAO sessionDao, IDatabase database) {
        LoginResult result = new LoginResult();

        try {
            Holder<String> errorMessage = new Holder<>();
            String condition = String.format("a.%s = %d", Session.Fields.user_sn, userNo);

            List<Map<String, Object>> dataList = joinDao.joinSessionUserGrade(condition, null, null, null, database, errorMessage);

            if (dataList == null) {
                System.err.println("Failed to read session");

                result.setSuccess(false);
                result.setMessage(ID.get(ErrorMessage.class, "failToReadSession").value());
                return result;
            }

            if (dataList.isEmpty()) {
                result.setSuccess(false);
                result.setMessage(ID.get(ErrorMessage.class, "noneSession").value());
                return result;
            }

            for (Map<String, Object> data : dataList) {
                Session session = (Session)data.get(Session.tableName);
                User user = (User)data.get(User.tableName);
                Grade grade = (Grade)data.get(Grade.tableName);

                if (session.getSession_key().equals(sessionKey)) {
                    if (!session.isAtmc_login_yn()) {
                        LocalDateTime sessionUpdate = session.getUpdt_de();
                        LocalDateTime now = LocalDateTime.now();

                        long diffSeconds = Duration.between(sessionUpdate, now).getSeconds();
                        if (diffSeconds > SESSION_SECONDS) {
                            result.setSuccess(false);
                            result.setMessage(ID.get(ErrorMessage.class, "logoutSession").value());
                            return result;
                        } else {
                            session.setUpdt_de(now);

                            if (!sessionDao.updateSession(session, errorMessage)) {
                                result.setSuccess(false);
                                result.setMessage(ID.get(ErrorMessage.class, "failToUpdateSession").value());
                                return result;
                            }
                        }
                    }

                    result.setSuccess(true);
                    result.setMessage(ID.get(ErrorMessage.class, "validSession").value());
                    result.setUser(ApplicationUser.makeUser(user, grade, session.getSession_key()));
                    return result;

                } else {
                    result.setSuccess(false);
                    result.setMessage(ID.get(ErrorMessage.class, "anotherOneLogin").value());
                    return result;
                }
            }

            result.setSuccess(false);
            result.setMessage(ID.get(ErrorMessage.class, "noneSession").value());

        } catch (Exception e) {
            result.setSuccess(false);
            result.setMessage(e.getMessage());
        }

        return result;
    }
}